weave/trace/weave_client.py

```diff
@@ -2111,4 +2196,71 @@ def elide_display_name(name: str) -> str:
     return name
 
 
-__docspec__ = [WeaveClient, Call, CallsIter]
+class ObjectVersionCollection:
+    """A collection of object versions for a single object.
+
+    This class provides access to all versions of a specific object.
+    By default, versions are sorted by version_index in ascending order
+    (oldest to newest), unless explicitly sorted differently when created.
+    """
+
+    def __init__(self, object_id: str, versions: list[ObjSchema], client: WeaveClient):
+        """Initialize a collection of object versions.
+
+        Args:
+            object_id: The ID of the object
+            versions: List of object versions (obj schemas)
+            client: WeaveClient instance for loading objects
+        """
+        self.object_id = object_id
+        self._raw_versions = sorted(versions, key=lambda v: v.version_index)
+        self._client = client
+        self._weave_objects: dict[int, Any] = {}  # {version: object}
+
+    def __iter__(self) -> Iterator[Any]:
+        for i in range(len(self._raw_versions)):
+            yield self._get_object_at_index(i)
+
+    def __len__(self) -> int:
+        return len(self._raw_versions)
+
+    def __getitem__(self, index: int) -> Any:
+        if index < 0:
+            # Technically it can be, but it will be confusing for most users.
+            raise IndexError(
+                "Negative indexing is not supported for ObjectVersionCollection"
+            )
+        elif index >= len(self._raw_versions):
+            raise IndexError(
+                f"Index {index} out of range for collection with {len(self._raw_versions)} versions"
+            )
+        return self._get_object_at_index(index)
+
+    def __repr__(self) -> str:
+        base_class = self.base_object_class or "Unknown"
+        return f"ObjectVersionCollection(object_id='{self.object_id}', base_class='{base_class}', versions={len(self._raw_versions)})"
+
+    def _get_object_at_index(self, index: int) -> Any:
+        if index not in self._weave_objects:
+            version = self._raw_versions[index]
+            entity, project = version.project_id.split("/", 1)
+
+            ref = ObjectRef(
+                entity=entity,
+                project=project,
+                name=version.object_id,
+                _digest=version.digest,
+            )
+            self._weave_objects[index] = self._client.get(ref)
+
+        return self._weave_objects[index]
+
+    @property
+    def base_object_class(self) -> str | None:
+        """Get the base object class of this object."""
+        if self._raw_versions:
+            return self._raw_versions[0].base_object_class
```

Comment:
One unfortunate thing is we allow different types of objects to be published with the same id. Worth considering how that impacts things like this.

Reply:
are you saying the unique key (and therefore the grouping) is `(base_object_class, object_id)`?

Reply:
I think you'd still want to group them by object_id alone, it is just that the base_object_class for the first version is not necessarily going to be the base_object_class for all versions.

Reply:
That seems bad -- when/why would the object class change across versions?

Reply:
You can reproduce this with just:
```
weave.publish(['felix', 'jimbo', 'billie'], 'same-name')
weave.publish({"foo": "bar"}, 'same-name')
weave.publish("hello", 'same-name')
```
As to why we allow it, possibly just an oversight in the original design, or possibly it was intentional to allow you to change the type of objects without losing the semantic connected-ness across versions of the same object, e.g. maybe you change a plain string used as a prompt to a StringPrompt object or something like that.